src/RunOpenCode/ExchangeRate/BancaIntesaSerbia/Source/ApiSource.php

```diff
@@ -0,0 +1,145 @@
+
+    /**
+     * @param Client $guzzleClient
+     */
+    public function __construct(Client $guzzleClient = null)
+    {
+        $this->guzzleClient = $guzzleClient !== null ? $guzzleClient : new Client();
+        $this->cache = [];
+    }
+
+    public function fetch($currencyCode, $rateType = RateType::MEDIAN, \DateTime $date = null)
+    {
+        $currencyCode = CurrencyCodeUtil::clean($currencyCode);
+
+        if (!Api::supports($currencyCode, $rateType)) {
+            throw new RuntimeException(sprintf('Banca Intesa Serbia does not supports currency code "%s" for rate type "%s".', $currencyCode, $rateType));
+        }
+
+        if ($date === null) {
+            $date = new \DateTime('now');
+        }
+
+        if (!\array_key_exists($rateType, $this->cache)) {
+            try {
+                $rates = $this->fetchRates($currencyCode, $date);
+                $this->cache[$currencyCode] = $rates;
+            } catch (\Exception $e) {
+                throw new SourceNotAvailableException(sprintf('Unable to load data from "%s" for "%s" of rate type "%s".', $this->getName(), $currencyCode, $rateType), 0, $e);
+            }
+        }
+
+        return $this->cache[$currencyCode][$rateType];
+    }
+
+    /**
+     * @return array<RateInterface>
+     */
+    private function fetchRates(string $currencyCode, \DateTime $date): array
+    {
+        $response = $this->guzzleClient->request('GET', self::URL, [
+            'query' => $this->getQueryParams($currencyCode, $date)
+        ]);
+
+        if (200 !== $response->getStatusCode()) {
+            throw new \Exception('Invalid response code, received "%s"', $response->getStatusCode());
+        }
+
+        $response = \json_decode($response->getBody()->getContents(), true);
+
+        if ('200' !== $response['statusCode']) { // For some reason, API returns HTTP status code 200, but sets the actual status code in the response body.. as a string..
+            throw new \Exception('Invalid response code, received "%s"', $response['statusCode']);
+        }
+
+        if (!\array_key_exists('rates', $response) || 0 === \count($response['rates'])) {
+            throw new \Exception('The rates for provided currency not received.');
+        }
+
+        $rates = $response['rates'][0]; // We always fetch result for a single day.
+        $apiRateTypes = $this->getApiRateTypes();
+
+        $extractedRates = [];
+
+        foreach (RateType::all() as $rateType) {
+            $extractedRates[$rateType] = $this->buildRate(
+                $rates[$apiRateTypes[$rateType]]['rate'],
+                $rates[$apiRateTypes[$rateType]]['currencyTo']['label'],
+                $rateType,
+                $date
+            );
+        }
+
+        return $extractedRates;
+    }
+
+    public function getName(): string
+    {
+        return Api::NAME;
+    }
+
+    private function buildRate($value, $currencyCode, $rateType, $date): RateInterface
```

Comment:
Do you need this private function? It is used once, and it is one liner.

Seams that you can just move construction to place where is needed, without using function, makes no sense to have it.